kordesii/utils/function_tracing/call_hooks/win_api/shell32.py

```diff
@@ -0,0 +1,114 @@
+        file_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+    ).decode("utf-16-le" if wide else "utf8")
+
+    if params_ptr:
+        params = cpu_context.read_data(
+            params_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+        ).decode("utf-16-le" if wide else "utf8")
+    else:
+        params = u""
+
+    if operation_ptr:
+        operation = cpu_context.read_data(
+            operation_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+        ).decode("utf-16-le" if wide else "utf8")
+    else:
+        operation = u""
+
+    if dir_ptr:
+        directory = cpu_context.read_data(
+            dir_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+        ).decode("utf-16-le" if wide else "utf8")
+    else:
+        directory = u""
+
+    logger.debug(f"{func_name}: {operation} {filepath}")
+    cpu_context.actions.append(
+        actions.ShellOperation(cpu_context.ip, operation, filepath, params, directory, visibility)
+    )
+
+    return random.randint(wc.MIN_HANDLE, wc.MAX_HANDLE)
+
+
+@builtin_func("SHGetFolderPathA")
+@builtin_func("SHGetFolderPathW")
+#typespec("SHFOLDERAPI SHGetFolderPathA(HWND hwnd, int    csidl,HANDLE hToken,DWORD  dwFlags,LPSTR  pszPath);")
+def sh_get_folder_path(cpu_context, func_name, func_args):
+    """
+    Retrieves the CSIDL folder mapping for the provided CSIDL Enumeration
+
+    Does not attempt to expand upon the folder path
+    """
+    wide = func_name.endswith("W")
+    _, csidl, _, _, path_ptr = func_args
+    csidl_int = construct.Int32ul.build(csidl)
+    try:
+        folder = construct.KnownFolderID(construct.Int32ul).parse(csidl_int)
+    except construct.ConstructError:
+        logger.warning(f"Could not acquire a folder for CSIDL {csidl}")
+        folder = f"{csidl}"
+    logger.debug(f"Writing CSIDL {folder} to 0x{path_ptr:08x}")
+    cpu_context.write_data(path_ptr, folder, data_type=constants.WIDE_STRING if wide else constants.STRING)
+    return wc.ERROR_SUCCESS
+
+
+@builtin_func("SHGetSpecialFolderPathA")
+@builtin_func("SHGetSpecialFolderPathW")
+#typespec("BOOL SHGetSpecialFolderPathA(HWND hwnd, LPSTR pszPath, int csidl, BOOL fCreate);")
+def sh_get_special_folder_path(cpu_context, func_name, func_args):
+    """
+    Retrieves the CSIDL folder mapping for the provided CSIDL Enumeration
+
+    Does not attempt to expand upon the folder path
+    """
+    wide = func_name.endswith("W")
+    _, path_ptr, csidl, _ = func_args
+    csidl_int = construct.Int32ul.build(csidl)
+    try:
+        folder = construct.KnownFolderID(construct.Int32ul).parse(csidl_int)
+    except construct.ConstructError:
+        logger.warning(f"Could not acquire a folder for CSIDL {csidl}")
+        folder = f"{csidl}"
+    logger.debug(f"Writing CSIDL {folder} to 0x{path_ptr:08x}")
+    cpu_context.write_data(path_ptr, folder, data_type=constants.WIDE_STRING if wide else constants.STRING)
+    return True
```

Comment:
This is an invalid data type returned and will be ignored by the emulator.
Function hooks should either return an integer, which will be set in rax, or None in order to not affect the rax register.

Should `wc.ERROR_SUCCESS` be returned instead?

Reply:
`SHGetSpecialFolderPath` explicitly returns a `BOOL` value, defined in [MSDN](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetspecialfolderpatha) as `True` or `False`, whereas `wc.ERROR_SUCCESS` is `0`, which would imply `False`. I can adjust this to return `1`, or can add to the `win_constants.py` to add `TRUE = 1` and `FALSE = 0`?

Reply:
The later option would be preferred.  Alternatively, you could update the `call` opcode to convert returned `True` and `False` values into `1` and `0` respectively.

Reply:
Updated the `call` opcode in 5f2c519 to convert the returned `True` and `False` values as described